flink-runtime/src/main/java/org/apache/flink/runtime/security/FlinkUserSecurityManager.java

```diff
@@ -0,0 +1,165 @@
+	 */
+	public static FlinkUserSecurityManager fromConfiguration(Configuration configuration) {
+		final String checkExitModeConfig = configuration.getString(SecurityOptions.CHECK_SYSTEM_EXIT);
+		final CheckExitMode checkExitMode;
+
+		try {
+			checkExitMode = CheckExitMode.valueOf(checkExitModeConfig.toUpperCase());
+		} catch (Exception ex) {
+			throw new IllegalConfigurationException(
+				String.format("%s is invalid configuration for %s.", checkExitModeConfig, SecurityOptions.CHECK_SYSTEM_EXIT.key()),
+				ex);
+		}
+
+		// If no check is enabled, return null so that caller can avoid setting security manager not to incur any runtime cost.
+		if (checkExitMode == CheckExitMode.DISABLED) {
+			return null;
+		}
+		// Add more configuration parameters that need user security manager (currently only for system exit).
+		return new FlinkUserSecurityManager(checkExitMode);
+	}
+
+	public static void setFromConfiguration(Configuration configuration) {
+		final FlinkUserSecurityManager flinkUserSecurityManager =
+			FlinkUserSecurityManager.fromConfiguration(configuration);
+		if (flinkUserSecurityManager != null) {
+			System.setSecurityManager(flinkUserSecurityManager);
+		}
+	}
+
+	public void checkPermission(Permission perm) {
+		if (originalSecurityManager != null) {
+			originalSecurityManager.checkPermission(perm);
+		}
+	}
+
+	public void checkPermission(Permission perm, Object context) {
+		if (originalSecurityManager != null) {
+			originalSecurityManager.checkPermission(perm, context);
+		}
+	}
+
+	public void checkExit(int status) {
+		super.checkExit(status);
+		if (!systemExitMonitored()) {
+			return;
+		}
+		switch (checkExitMode) {
+			case DISABLED:
+				break;
+			case WARN:
+				// Add exception trace log to help users to debug where exit came from.
+				LOG.warn("Exiting JVM with status {} is monitored, logging and exiting",
+					status, new UserSystemExitException());
+				break;
+			case THROW:
+				throw new UserSystemExitException();
+			default:
+				// Must not happen if exhaustively handling all modes above. Logging as being already at exit path.
+				LOG.warn("No valid check exit mode configured: {}", checkExitMode);
+		}
+	}
+
+	public SecurityManager getOriginalSecurityManager() {
+		return originalSecurityManager;
+	}
+
+	public void monitorSystemExit() {
+		monitorSystemExit.set(true);
+	}
+
+	public void unmonitorSystemExit() {
+		monitorSystemExit.set(false);
+	}
+
+	public boolean systemExitMonitored() {
+		return Boolean.TRUE.equals(monitorSystemExit.get());
+	}
+
+	public static void monitorSystemExitForCurrentThread() {
+		SecurityManager securityManager = System.getSecurityManager();
+		if (securityManager instanceof  FlinkUserSecurityManager) {
+			((FlinkUserSecurityManager) securityManager).monitorSystemExit();
+		}
+	}
+
+	public static void unmonitorSystemExitForCurrentThread() {
+		SecurityManager securityManager = System.getSecurityManager();
```

Comment:
Great point. The original patch was written long ago before while no exit security manager existed. I think we should consider this correctly. There would be two ways: 1) forcing FlinkUserSecurityManager at the end, or 2) making this manager as a singleton, so that it could be configurable regardless of where it's chained. I will take a look at how it looks like.